via_verifier/lib/via_musig2/src/utxo_manager.rs

```diff
@@ -0,0 +1,426 @@
+use std::{collections::VecDeque, sync::Arc};
+
+use bitcoin::{Address, Amount, OutPoint, Transaction, TxOut};
+use tokio::sync::RwLock;
+use via_btc_client::traits::BitcoinOps;
+
+const CTX_REQUIRED_CONFIRMATIONS: u32 = 1;
+const DEFAULT_CAPACITY: usize = 100;
+
+#[derive(Debug, Clone)]
+pub struct UtxoManager {
+    /// Btc client
+    btc_client: Arc<dyn BitcoinOps>,
+    /// The wallet address
+    address: Address,
+    /// The transactions executed by the wallet
+    context: Arc<RwLock<VecDeque<Transaction>>>,
+    /// The minimum amount to merge utxos
+    minimum_amount: Amount,
+    /// The maximum number of utxos to merge in a single tx
+    merge_limit: usize,
+}
+
+impl UtxoManager {
+    pub fn new(
+        btc_client: Arc<dyn BitcoinOps>,
+        address: Address,
+        minimum_amount: Amount,
+        merge_limit: usize,
+    ) -> Self {
+        UtxoManager {
+            btc_client,
+            address,
+            context: Arc::new(RwLock::new(VecDeque::with_capacity(DEFAULT_CAPACITY))),
+            minimum_amount,
+            merge_limit,
+        }
+    }
+
+    pub async fn get_available_utxos(&self) -> anyhow::Result<Vec<(OutPoint, TxOut)>> {
+        // fetch utxos from client
+        let mut utxos = self.btc_client.fetch_utxos(&self.address).await?;
+        {
+            if self.context.read().await.is_empty() {
+                return Ok(utxos);
+            }
+        }
+
+        // Add the output utxos to the list
+        for tx in self.context.read().await.iter() {
+            for (i, out) in tx.output.iter().enumerate() {
+                if out.script_pubkey != self.address.script_pubkey() {
+                    continue;
+                };
+
+                let outpoit = OutPoint {
+                    txid: tx.compute_txid(),
+                    vout: i as u32,
+                };
+                utxos.push((outpoit, tx.output[i].clone()));
+            }
+        }
+
+        // Remove the inputs used utxos
+        for tx in self.context.read().await.iter() {
+            for input in tx.input.iter() {
+                let outpoint = input.previous_output;
+                let index = utxos.iter().position(|(op, _)| op == &outpoint);
+                if let Some(index) = index {
+                    utxos.remove(index);
+                }
+            }
+        }
```

Comment:
Lets try to optimize this by reducing the number of times the read lock is acquire and combining the 2 loops into 1.

```suggestion
        let context = self.context.read().await;

        {
            if context.is_empty() {
                return Ok(utxos);
            }
        }

        for tx in context.iter() {
            // Add the output utxos to the list
            for (i, out) in tx.output.iter().enumerate() {
                if out.script_pubkey == self.address.script_pubkey() {
                    let outpoint = OutPoint {
                        txid: tx.compute_txid(),
                        vout: i as u32,
                    };
                    utxos.push((outpoint, out.clone()));
                }
            }
            
            // Remove the inputs used utxos
            for input in &tx.input {
                if let Some(index) = utxos.iter().position(|(op, _)| op == &input.previous_output) {
                    utxos.remove(index);
                }
            }
        }
```